### DISCUSSION TITLE
Difference rowcount behavior between psycopg2 and psycopg, not documented

### DISCUSSION THREAD

**Discussion by CaselIT** [psycopg/psycopg#732] [General]
Hi,

I've noticed a difference in how the rowcount of a closed cursor works between psycopg2 and psycopg: in psycopg2 a closed cursor kept the last query rowcount, with psyocpg resets it to -1.
The following pattern is basically what sqlalchemy does when running inserts (it closes the cursor before returning the result if it does not returns rows)
```py
import psycopg
import psycopg2

url = "postgresql://scott:tiger@localhost/test"

def using_driver(mod):
    if mod == "psycopg2":
        con = psycopg2.connect(url)
    else:
        con = psycopg.connect(url)
    with con:
        with con.cursor() as c:
            c.execute("create table tbl (col int)")
        with con.cursor() as c:
            c.execute("insert into tbl values (1), (2), (3)")
            print(mod, "insert values", c.rowcount)
        print(mod, "insert values", c.rowcount)
        with con.cursor() as c:
            c.execute("insert into tbl values (1), (2), (3) returning 1")
        print(mod, "insert values returning", c.rowcount)
        with con.cursor() as c:
            c.execute(
                "insert into tbl select t from "
                "generate_series(%s::int,%s::int) as t ",
                (5, 9),
            )
        print(mod, "insert select params", c.rowcount)
        with con.cursor() as c:
            c.executemany("insert into tbl values (%s)", [(8,), (9,)])
        print(mod, "insert values exec many", c.rowcount)
        con.rollback()

using_driver("psycopg2")
using_driver("psycopg")
```

The above code prints
```
psycopg2 insert values 3
psycopg2 insert values returning 3
psycopg2 insert select params 5
psycopg2 insert values exec many 2
psycopg insert values -1
psycopg insert values returning -1
psycopg insert select params -1
psycopg insert values exec many -1
```

Is this by design? 
If it may be worth mentioning it in https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html ?
If not is it reasonable to restore the psycopg2 behaviour?

I kinda think it's by design but I though it's worth asking.

I've noticed this while updated a work project from psycopg2 to psycopg3 in sqlalchemy. The issue there is the following: https://github.com/sqlalchemy/sqlalchemy/issues/10974

My use case in the project is getting the number of rows inserted in an `insert ... select ...`, where otherwise there is no easy way of getting how many rows the query would match. My current workaround is to add `returning 1` and then count the number of rows that it returned, but that's less than ideal

**Comment by dvarrazzo** [comment]
Hello, Federico,

after it is closed, the cursor is unusable. What you are seeing in psycopg2 is likely an implementation accident, with the value copied to the C structure and not reset on `close()`. Psycopg 3 has a cleaner cleanup function (invoked between `execute()` calls, to prepare the cursor to the new results, and on `close()` to free up resources), and `rowcount` is reset there too, consistently with the rest of the state, and I would say more correctly than in psycopg2.

I don't think it is worth mentioning it in the porting guide because, neither in psycopg 2 nor in 3, you should be using the cursor after closing. I don't want to commit to an interface the the behaviour of a closed cursor.

> My use case in the project is getting the number of rows inserted in an `insert ... select ...`, where otherwise there is no easy way of getting how many rows the query would match.

In your use case, what stops you from reading `rowcount` before closing the cursor?

**Reply by CaselIT** [reply]
I kinda suspected it was by design. I also agree that it's a better interface.

> In your use case, what stops you from reading rowcount before closing the cursor?

I'm using sqlalchemy and the cursor closure is hardcoded before a result is returned to the user in case of a statement that doesn't return rows.

I'll see if there are workaround on sqlalchemy side.

Thanks for the reply!
